docker/stan-wasm-server/logic/compilation.py

```diff
@@ -38,60 +37,46 @@ def copy_compilation_outputs(*, model_dir: Path, job_dir: Path):
     for (src, dest) in todos:
         copy2(src, dest)
 
+async def compile_and_cache(*, job_id: str, model_dir: Path, tinystan_dir: Path):
+
+    # if there's a cache hit, make sure any copying is already complete,
+    # then return
+    if all((model_dir / x).exists() for x in COMPILATION_OUTPUTS):
+        wait_until_free(model_dir)
+        return
```

Comment:
Are we doing the "check for all compilation outputs" thing elsewhere? (e.g. the possibly-removed `copy_compilation_outputs` function above)

If so, could potentially standardize the file check.

Reply:
Not currently

Reply:
I realized another condition which could lead to an error, so this is now pulled out.

The issue was another sequencing one:

```
THREAD A starts compiling
THREAD B starts compiling
THREAD A finishes
THREAD A acquires the lock
THREAD A copies the files
THREAD A releases the lock

THREAD B finishes
THREAD B acquires the lock
THREAD B begins copying, WHILE THREAD A requests to download the files

THREAD A reads partially completed copies
```
This is fixed by having THREAD B check if the files exist before copying them, and I've tried to explain it in a comment in the file

Reply:
Yep, want to check for existence both before and after getting the lock. Good call.